### PR TITLE
all: pluggable consensus engines

### PR DESCRIPTION
**Pull request by karalabe**
The PR implements a pluggable consensus engine for go-ethereum. In short, it introduces a generic consensus interface, and refactors the entire codebase to use this interface:

```go
// Engine is an algorithm agnostic consensus engine.
type Engine interface {
	// VerifyHeader checks whether a header conforms to the consensus rules of a
	// given engine. Verifying the seal may be done optionally here, or explicitly
	// via the VerifySeal method.
	VerifyHeader(chain ChainReader, header *types.Header, seal bool) error

	// VerifyHeaders is similar to VerifyHeader, but verifies a batch of headers
	// concurrently. The method returns a quit channel to abort the operations and
	// a results channel to retrieve the async verifications (the order is that of
	// the input slice).
	VerifyHeaders(chain ChainReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error)

	// VerifyUncles verifies that the given block's uncles conform to the consensus
	// rules of a given engine.
	VerifyUncles(chain ChainReader, block *types.Block) error

	// VerifySeal checks whether the crypto seal on a header is valid according to
	// the consensus rules of the given engine.
	VerifySeal(chain ChainReader, header *types.Header) error

	// Prepare initializes the consensus fields of a block header according to the
	// rules of a particular engine. The changes are executed inline.
	Prepare(chain ChainReader, header *types.Header) error

	// Finalize runs any post-transaction state modifications (e.g. block rewards)
	// and assembles the final block.
	//
	// Note, the block header and state database might be updated to reflect any
	// consensus rules that happen at finalization (e.g. block rewards).
	Finalize(chain ChainReader, header *types.Header, state *state.StateDB, txs []*types.Transaction,
		uncles []*types.Header, receipts []*types.Receipt) (*types.Block, error)

	// Seal generates a new block for the given input block with the local miner's
	// seal place on top.
	Seal(chain ChainReader, block *types.Block, stop <-chan struct{}) (*types.Block, error)

	// APIs returns the RPC APIs this consensus engine provides.
	APIs(chain ChainReader) []rpc.API
}
```

This PR does only the ethash refactors to fit on top of this, but the interface and related code was done in such a way as to fit clique (poa) on top too, which you can find in https://github.com/ethereum/go-ethereum/pull/3753.

As a side modification, PoW threading is also handled now by `ethash`, and not by the miner, which moves a lot of messy threading logic out of there and makes things much more robust.

### CODE REVIEW

File: mobile/params.go [hunk 1]
```diff
@@ -71,14 +73,25 @@ func TestnetGenesis() string {
...
+	EIP150Block    int64      // Homestead gas reprice switch block
+	EIP150Hash     Hash       // Homestead gas reprice switch block hash
+	EIP155Block    int64      // Replay protection switch block
+	EIP158Block    int64      // Empty account pruning switch block
+	PoWConfig      *PoWConfig // Consensus rules for the proof-of-work engine
+	PoAConfig      *PoAConfig // Consensus rules for the proof-of-authority engine
+}
+
+// PoWConfig is the consensus engine configs for proof-of-work based sealing.
+type PoWConfig struct{}
```
Comment (zsfelfoldi) [inline mobile/params.go]: Why is this necessary?

File: consensus/utils.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+type CurrentHeaderFn func() *types.Header
+type HeaderRetrievalFn func(hash common.Hash, number uint64) *types.Header
+type HeaderByNumberRetrievalFn func(number uint64) *types.Header
+type BlockRetrievalFn func(hash common.Hash, number uint64) *types.Block
+
+// hookedChainReader is a tiny implementation of ChainReader based on callbacks
+// and a preset chain configuration. It's useful to avoid defining various custom
+// types where a cain reader isn't readily available.
+type hookedChainReader struct {
```
Comment (zsfelfoldi) [inline consensus/utils.go]: Why? MakeChainReader is not even used at the moment, and whatever would use it in the future could just as well directly implement ChainReader.

Comment (karalabe) [inline consensus/utils.go]: It's a leftover from a previous prototype. I need to cealn it out.

File: mobile/params.go (hunk 1 above)
Comment (karalabe) [inline mobile/params.go]: It's not in this pr. The poa will need a variation of this. I need to move it out from this commit.

File: miner/agent.go [hunk 3]
```diff
@@ -17,56 +17,49 @@
...
 	return miner
 }
 
 func (self *CpuAgent) Work() chan<- *Work            { return self.workCh }
-func (self *CpuAgent) Pow() pow.PoW                  { return self.pow }
 func (self *CpuAgent) SetReturnCh(ch chan<- *Result) { self.returnCh = ch }
 
 func (self *CpuAgent) Stop() {
-	close(self.quit)
+	self.stop <- struct{}{}
```
Comment (zsfelfoldi) [inline miner/agent.go]: Should be OK but why is it better than simply closing the channel?

Comment (karalabe) [inline miner/agent.go]: The CPU agent is reused between works. Recreating channels is very dangerous as it's extremely easy to produce data races. Previously we created new miners wily nilly (exactly because it wasn't safe to recreate the channel alone within a miner), which lead to a lot of issues. By moving threading into `ethash`, the mining code is not clean enough not to require constant recreations.

File: core/block_validator.go [hunk 4]
```diff
@@ -44,30 +34,24 @@ var (
...
-// ValidateBlock also validates and makes sure that any previous state (or present)
-// state that might or might not be present is checked to make sure that fast
-// sync has done it's job proper. This prevents the block validator from accepting
-// false positives where a header is present but the state is not.
-func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+// ValidateBody validates the given block's uncles and verifies the the block
+// header's transaction and uncle roots. The headers are assumed to be already
+// validated at this point.
+func (v *BlockValidator) ValidateBody(block *types.Block) error {
+	// Check whether the blokc's known, and if not, that it's linkable
```
Comment (Gustav-Simonsson) [inline core/block_validator.go]: block

File: consensus/ethash/consensus.go [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+	}
+	// If we're running a shared PoW, delegate verification to it
+	if ethash.shared != nil {
+		return ethash.shared.VerifySeal(chain, header)
+	}
+	// Sanity check that the block number is below the lookup table size (60M blocks)
+	number := header.Number.Uint64()
+	if number/epochLength >= uint64(len(cacheSizes)) {
+		// Go < 1.7 cannot calculate new cache/dataset sizes (no fast prime check)
+		return ErrNonceOutOfRange
```
Comment (holiman) [inline consensus/ethash/consensus.go]: So.. this won't be a problem until after 60M blocks - or does the comment mean something else?

File: consensus/ethash/consensus.go [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+				case chain.GetHeader(headers[index].Hash(), headers[index].Number.Uint64()-1) != nil:
+					outputs <- result{index: index, err: nil}
+				case parent == nil:
+					outputs <- result{index: index, err: ErrParentUnknown}
+				default:
+					outputs <- result{index: index, err: ethash.verifyHeader(chain, headers[index], parent, false, seals[index])}
+				}
+			}
+		}()
+	}
```
Comment (holiman) [inline consensus/ethash/consensus.go]: If there's a chain of headers, and one of them does not validate - will this continue to validate the rest of the chain, or will it abort at the earliest chance? It seems to me it will keep validating (uselessly)... ? 

If so, an attacker could feed long sequences of invalid chains... ?

File: miner/remote_agent.go [hunk 7]
```diff
@@ -129,8 +131,8 @@ func (a *RemoteAgent) GetWork() ([3]string, error) {
 	return res, errors.New("No work available yet, don't panic.")
 }
 
-// SubmitWork tries to inject a PoW solution tinto the remote agent, returning
-// whether the solution was acceted or not (not can be both a bad PoW as well as
+// SubmitWork tries to inject a pow solution tinto the remote agent, returning
```
Comment (Gustav-Simonsson) [inline miner/remote_agent.go]: into, accepted

File: core/block_validator.go (hunk 4 above)
Comment (karalabe) [inline core/block_validator.go]: I missed you too <3

Comment (karalabe) [inline core/block_validator.go]: :D

File: consensus/ethash/consensus.go (hunk 6 above)
Comment (karalabe) [inline consensus/ethash/consensus.go]: Good point. Though we did do it like this until now too, just not for full header validation, only for PoW validation. https://github.com/ethereum/go-ethereum/blob/master/core/chain_pow.go#L65

Does validating more headers (maximum batch size is 2K I think, probably less usually when syncing) have a measurable impact on performance? Given that it's just a few extra calculations (difficulty, gas limit), it may not be worth the complexity to support early terminations.

File: consensus/ethash/consensus.go [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+		return ErrTooManyUncles
+	}
+	// Gather the set of past uncles and ancestors
+	uncles, ancestors := set.New(), make(map[common.Hash]*types.Header)
+
+	number, parent := block.NumberU64()-1, block.ParentHash()
+	for i := 0; i < 7; i++ {
+		ancestor := chain.GetBlock(parent, number)
+		if ancestor == nil {
+			break
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: I didn't write this code (just copied over from core), so I can't say for certain, but I'd assume it's for cases when you are importing blocks at the beginning of the chain where you don't have 7 blocks yet.

File: consensus/ethash/consensus.go (hunk 5 above)
Comment (karalabe) [inline consensus/ethash/consensus.go]: Yes. But that's how it works currently too. https://github.com/ethereum/ethash/blob/master/ethash.go#L133

File: consensus/ethash/sealer.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	ethash.lock.Unlock()
+	if threads == 0 {
+		threads = runtime.NumCPU()
+	}
+	var pend sync.WaitGroup
+	for i := 0; i < threads; i++ {
+		pend.Add(1)
+		go func(id int) {
+			defer pend.Done()
+			ethash.mine(block, id, uint64(rand.Int63()), abort, found)
```
Comment (karalabe) [inline consensus/ethash/sealer.go]: I don't see why that would be the case. `math/rand` only makes subsequent calls correlated in some way, but I don't see how multiple nodes in the network could end up with the same seed state at the same instant in time.

File: core/block_validator_test.go [hunk 10]
```diff
@@ -17,64 +17,173 @@
...
+	for i, block := range blocks {
+		headers[i] = block.Header()
+		seals[i] = true
 	}
+	// Set the number of threads to verify on
+	old := runtime.GOMAXPROCS(threads)
+	defer runtime.GOMAXPROCS(old)
+
+	// Run the header checker for the entire block chain at once both for a valid and
+	// also an invalid chain (enough if one is invalid, last but one (arbitrary)).
```
Comment (karalabe) [inline core/block_validator_test.go]: `last-but-one`

It doesn't matter which block is invalid, so I chose one from the end, last-but-one. :) I agree the formulation is catastrophic.

File: consensus/ethash/sealer.go (hunk 9 above)
Comment (fjl) [inline consensus/ethash/sealer.go]: It depends on whether the global math/rand instance is seeded or not. It would be better so seed a separate instance for the purpose of generating nonces.

File: consensus/ethash/consensus.go (hunk 6 above)
Comment (holiman) [inline consensus/ethash/consensus.go]: I suggest to use a simple shortcut in the event an invalid block has been found, sth like https://gist.github.com/holiman/d2460e0814ad0fb12135ca4e7764da0d

**Comment by karalabe** [comment]
@zsfelfoldi Removed all the dead code you spotted, thanks. PTAL
@holiman Added early terminations to ethash header verification if any worker finds a bad block. The block is shared between all workers via atomics, so any worker can spin all others off too. PTAL.
@Gustav-Simonsson @fjl I've added a random source to each ethash instance that is initialized from `crypto/rand` the first time the miner starts up. This should ensure that the package cannot be meaningfully misused or abused. PTAL.
@Gustav-Simonsson Addressed all typos. Thanks for them!

**Comment by holiman** [comment]
:+1: Nice work!

File: miner/worker.go [hunk 11]
```diff
@@ -424,23 +417,29 @@ func (self *worker) commitNewWork() {
 		tstamp = parent.Time().Int64() + 1
 	}
 	// this will ensure we're not going off too far in the future
-	if now := time.Now().Unix(); tstamp > now+4 {
+	if now := time.Now().Unix(); tstamp > now+1 {
```
Comment (karalabe) [inline miner/worker.go]: @holiman Btw, your question here still stands. I'm not sure why this original 4 second into-the-future thing was allowed in the first place. My "fixed" code just ensures that future blocks cannot be minted. 

I'm guessing it was meant for the Frontier launch (or testnets) when difficulty was low and miners found multiple blocks a second. However I think it's an ugly and unnecessary hack really to allow future blocks.

Comment (karalabe) [inline miner/worker.go]: There's also a drawback, that if the miner "sleeps" for 4 seconds (which happened previously), but blocks keep pouring in, a lot of weird things could happen.

We should really rewrite the miner.

File: consensus/ethash/consensus.go [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+		}
+	}()
+	return abort, returns
+}
+
+// VerifyUncles verifies that the given block's uncles conform to the consensus
+// rules of the stock Ethereum ethash engine.
+func (ethash *Ethash) VerifyUncles(chain consensus.ChainReader, block *types.Block) error {
+	// If we're running a full engine faking, accept any input as valid
+	if ethash.fakeFull {
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: There are different variations of fakeness. Some only fake out the PoW, others the entire consensus rules (such as difficulty), yet some only fail certain blocks. Doing only fake-pow but true consensus checks would become really messy if done in a new object.

File: consensus/ethash/consensus.go [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+		return nil
+	}
+	// Verify that there are at most 2 uncles included in this block
+	if len(block.Uncles()) > 2 {
```
Comment (karalabe) [inline consensus/ethash/consensus.go]: I was also thinking about this. Though perhaps I'd put it into a params.go inside the `ethash` package, since uncles are only meaningful for ethash, and not for anything else (clique currently, but probably casper too).

File: core/types/block.go [hunk 14]
```diff
@@ -120,6 +120,33 @@ func (h *Header) HashNoNonce() common.Hash {
 	})
 }
 
+// HashNoPoASig returns the hash which is used as input for the proof-of-authority
```
Comment (karalabe) [inline core/types/block.go]: No. Felix made me split my clique PR up even though I told him it's getting messy since it's hard to find all the intermingled code snippets that are only needed by clique. I made quite a few "updates" already to move snippets out that Zsolt and others found.

I'm reluctant to rebase all 3 of my PRs once again to move this to the second commit. Can do if really necessary.

File: core/block_validator.go [hunk 15]
```diff
@@ -80,30 +64,17 @@ func (v *BlockValidator) ValidateBlock(block *types.Block) error {
 	if _, err := state.New(parent.Root(), v.bc.chainDb); err != nil {
 		return ParentError(block.ParentHash())
 	}
-
+	// Header validity is known at this point, check the uncles and transactions
```
Comment (karalabe) [inline core/block_validator.go]: When importing blocks or headers, previously we only checked PoW, and the block validator did everything else. As of this PR, the `consensus.Engine` checks both PoW as well as header consensus fields, since they are different depending on whether we use `ethash` or `clique`. Block contents are the same, so I left them unchanged.

File: consensus/ethash/consensus.go (hunk 12 above)
Comment (Arachnid) [inline consensus/ethash/consensus.go]: Would it? Several of the methods are just shortcut with `if ethash.fakeFull`; couldn't you provide an implementation that either embeds the regular ethash and forwards only some methods, or have both implementations call the same underlying functions?

File: consensus/ethash/consensus.go (hunk 13 above)
Comment (Arachnid) [inline consensus/ethash/consensus.go]: Sounds good. Magic numbers bad.

File: consensus/ethash/consensus.go (hunk 12 above)
Comment (fjl) [inline consensus/ethash/consensus.go]: I agree with @Arachnid.

Comment (karalabe) [inline consensus/ethash/consensus.go]: I'll do the changes on Tuesday when I get back. Please review the remainder until now so we can merge afterwards.

File: core/types/block.go (hunk 14 above)
Comment (karalabe) [inline core/types/block.go]: I'll change this too on Tuesday.

File: consensus/ethash/consensus.go (hunk 12 above)
Comment (karalabe) [inline consensus/ethash/consensus.go]: There are more intricate interplays where the checks whether it's a tester ethash is deep inside the code, not just an outer check: https://github.com/ethereum/go-ethereum/pull/3817/files#diff-ba80760e0df82b1a7db2323760cc574bR472 (among other similar ones). This would mean that I would either need to reimplement this entire function for the testers, or would need to add some sort of hook, which bings us back to the original code.

File: core/types/block.go (hunk 14 above)
Comment (karalabe) [inline core/types/block.go]: Fixed.

File: consensus/ethash/consensus.go (hunk 13 above)
Comment (karalabe) [inline consensus/ethash/consensus.go]: Fixed.

**Comment by karalabe** [comment]
So, we kind of agreed off github that the package moves won't be done at the moment. Mostly because it cannot be done fully clean. We might try again some time later when more consensus engines are added and we see what other stuff we can clean up in core.

As for the `fake` wrappers. I had a go at it but realized soon enough that it's actually not doable cleanly. The issue is that PoW is validated deep inside ethash, after other consensus rules pass. So I can easily enough wrap outer calls with fake checks, but to fake out PoW, I need to hook deep into the real ethash implementation to make it call the fake checks (this is a case where inheritance would be able to solve the issue, since even the real ethash would call the outer VerifySeal).

Long story short, I don't think I have any further modifications to make at the moment. @Arachnid @fjl PTAL
